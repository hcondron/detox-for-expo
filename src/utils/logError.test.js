const { exec } = require('child-process-promise');
const DetoxRuntimeError = require('../errors/DetoxRuntimeError');
const mockLogger = require('./__mocks__/logger');
const logError = require('./logError');

describe('logError', () => {
  it('should not fail on nulls', () => {
    logError(mockLogger, null);
    expect(mockLogger.error.mock.calls).toMatchSnapshot();
  });

  it('should skip child process errors', async () => {
    await exec('sdfhkdshfjksdhfjkhks').catch((e) => logError(mockLogger, e));
    expect(mockLogger.error).not.toHaveBeenCalled();
    expect(mockLogger.warn).not.toHaveBeenCalled();
  });

  it('should log long detox runtime errors', () => {
    const err = new DetoxRuntimeError({
      message: 'msg123',
      hint: 'hint',
      debugInfo: 'debugInfo',
    });
    hideStack(err);

    logError(mockLogger, err);

    expect(mockLogger.error.mock.calls).toMatchSnapshot();
    expect(mockLogger.warn.mock.calls).toMatchSnapshot();
  });

  it('should log short detox runtime errors', () => {
    const err = new DetoxRuntimeError({ message: 'short' });
    hideStack(err);
    logError(mockLogger, err);

    expect(mockLogger.error.mock.calls).toMatchSnapshot();
    expect(mockLogger.warn).not.toHaveBeenCalled();
  });

  it('should log other errors', () => {
    const err = new Error('message');
    hideStack(err);
    logError(mockLogger, err);

    expect(mockLogger.error).toHaveBeenCalledWith({ err }, err);
    expect(mockLogger.warn).not.toHaveBeenCalled();
  });
});

function hideStack(err) {
  err.stack = replaceLinesAfterTheFirst(err.stack, '  at fake error stack');
}

function replaceLinesAfterTheFirst(str, replacement) {
  const [first, ...rest] = str.split('\n');

  return (rest.length > 0)
    ? first + '\n' + replacement
    : first;
}
