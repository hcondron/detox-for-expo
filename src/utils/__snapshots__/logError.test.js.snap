// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logError should log long detox runtime errors 1`] = `
Array [
  Array [
    Object {
      "stack": true,
    },
    "DetoxRuntimeError: DetoxRuntimeError: msg123
  at fake error stack",
  ],
]
`;

exports[`logError should log long detox runtime errors 2`] = `
Array [
  Array [
    Object {
      "hint": true,
    },
    "Hint: hint",
  ],
  Array [
    Object {
      "debugInfo": true,
    },
    "See debug info below:
debugInfo",
  ],
]
`;

exports[`logError should log short detox runtime errors 1`] = `
Array [
  Array [
    Object {
      "stack": true,
    },
    "DetoxRuntimeError: DetoxRuntimeError: short
  at fake error stack",
  ],
]
`;

exports[`logError should not fail on nulls 1`] = `
Array [
  Array [
    "Unknown error (null)",
  ],
]
`;
